src/constructors/mycielskian.c

```diff
@@ -0,0 +1,253 @@
+        IGRAPH_CHECK(igraph_vector_int_push_back(&edges, 1));
+    }
+
+    if (vcount == 1) { // single vertex, assuming no self loop
+        if (k == 0) {
+            IGRAPH_CHECK(igraph_empty(res, 1, IGRAPH_UNDIRECTED));
+            IGRAPH_FINALLY_CLEAN(1);
+            igraph_vector_int_destroy(&edges);
+            return IGRAPH_SUCCESS;
+        }
+        /* else */
+        init_iter = 1; // start from 1st iteration
+        // Make a 2-vertex graph
+        vcount = 2;
+        ecount = 1;
+        IGRAPH_CHECK(igraph_vector_int_push_back(&edges, 0));
+        IGRAPH_CHECK(igraph_vector_int_push_back(&edges, 1));
+    }
+
+    igraph_integer_t new_vcount = vcount;
+    igraph_integer_t new_ecount = ecount;
+
+    for (igraph_integer_t i = init_iter; i < k; i++) {
+        // new edges = 3 * old edges + old vertices
+        IGRAPH_SAFE_MULT(new_ecount, 3, &new_ecount);
+        IGRAPH_SAFE_ADD(new_ecount, new_vcount, &new_ecount);
+
+        // new vertices = 2 * old vertices + 1
+        IGRAPH_SAFE_MULT(new_vcount, 2, &new_vcount);
+        IGRAPH_SAFE_ADD(new_vcount, 1, &new_vcount);
+    }
+
+    // copy the edges from the original graph to the new vector
+    // excluding for the case of vcount==0 and vcount==1
+    // i.e, init_iter != 0, exclude it
+    if (init_iter == 0) {
+        IGRAPH_CHECK(igraph_get_edgelist(graph, &edges, false));
+    }
+    IGRAPH_CHECK(igraph_vector_int_resize(&edges, new_ecount * 2));
+
+    igraph_integer_t edge_index = 2 * ecount;  // Current last edge index in edge vector
+    igraph_integer_t offset = vcount;          // Tracks where new vertices start
+
+    for (igraph_integer_t i = init_iter; i < k; i++) {
+        igraph_integer_t prev_vcount = offset;  // Number of vertices before this step
+        igraph_integer_t w = offset * 2;        // The new 'w' node index
+        igraph_integer_t last_edge_index = edge_index;  // Mark where edges before this step end
+
+        // For each edge before this step, add two new edges
+        for (igraph_integer_t j = 0; j < last_edge_index; j += 2) {
+            igraph_integer_t v1 = VECTOR(edges)[j];
+            igraph_integer_t v2 = VECTOR(edges)[j + 1];
+
+            VECTOR(edges)[edge_index++] = v1;
+            VECTOR(edges)[edge_index++] = offset + v2;
+
+            VECTOR(edges)[edge_index++] = v2;
+            VECTOR(edges)[edge_index++] = offset + v1;
+        }
+
+        // Add edges connecting each `ui` to `w` (forming a star)
+        for (igraph_integer_t j = prev_vcount; j < w; j++) {
+            VECTOR(edges)[edge_index++] = j;
+            VECTOR(edges)[edge_index++] = w;
+        }
+
+        // Update offset for next step
+        offset = offset * 2 + 1;
+
+        IGRAPH_ALLOW_INTERRUPTION();
+    }
+
+    // Add all edges in one go
+    IGRAPH_CHECK(igraph_create(res, &edges, 0, igraph_is_directed(graph)));
+    IGRAPH_FINALLY_CLEAN(1);
+    igraph_vector_int_destroy(&edges);
```

Comment:
For clarity, `IGRAPH_FINALLY_CLEAN()` should always come after the destructors. I'll fix this.